actors/market/src/lib.rs

```diff
@@ -1025,6 +1000,74 @@ impl Actor {
             }
         }
     }
+
+    fn process_deal_updates(
+        rt: &impl Runtime,
+        params: ProcessDealsParams,
+    ) -> Result<ProcessDealsReturn, ActorError> {
+        rt.validate_immediate_caller_accept_any()?;
+        let curr_epoch = rt.curr_epoch();
+        let mut total_slashed = TokenAmount::zero();
+        let mut expired_deals: Vec<DealID> = Vec::new();
+        let mut terminated_deals: Vec<DealID> = Vec::new();
+
+        rt.transaction(|st: &mut State, rt| {
+            let mut new_deal_states: Vec<(DealID, DealState)> = Vec::new();
+            for deal_id in params.deal_ids {
+                let deal_proposal = st.get_proposal(rt.store(), deal_id)?;
+                let dcid = rt_deal_cid(rt, &deal_proposal)?;
+
+                let (deal_state, slashed) = st.get_active_deal_or_process_timeout(
+                    rt.store(),
+                    curr_epoch,
+                    deal_id,
+                    &deal_proposal,
+                    &dcid,
+                )?;
+
+                let mut deal_state = match deal_state {
+                    Some(deal_state) => deal_state,
+                    None => {
+                        // deal was cleaned up
+                        total_slashed += slashed;
+                        expired_deals.push(deal_id);
+                        continue;
+                    }
+                };
+
+                let (slash_amount, remove_deal) = st.process_deal_update(
+                    rt.store(),
+                    &deal_state,
+                    &deal_proposal,
+                    &dcid,
+                    curr_epoch,
+                )?;
+
+                if remove_deal {
+                    total_slashed += slash_amount;
+                    st.remove_completed_deal(rt.store(), deal_id)?;
+                    terminated_deals.push(deal_id);
```

Comment:
Please document how it's possible to get here after code upgrade, given that the code for syncrhonous termination will appear to make it impossible.

Reply:
Oh, I think this mixes terminated (temporary case around teh upgrade) with merely completed. Please separate them. You might need another flag in the return value from `st.process_deal_update` to distinguish the cases.

Reply:
processing slashed deals here is now an error so terminated deals are no longer handled